room: move flipmap vars and logic to TRX

lahm86 opened the pull request:
#### Checklist

- [x] I have read the [coding conventions](https://github.com/LostArtefacts/TRX/blob/master/CONTRIBUTING.md#coding-conventions)
- [x] I have added a changelog entry about what my pull request accomplishes, or it is an internal change
- [x] I have added a readme entry about my new feature or OG bug fix, or it is a different change

#### Description

This moves all variables related to flipmaps to TRX, along with the main logic to flip the map. We also had a comment previously in `Room_FlipMap` to check if item/effect assignment was required; I tested this and it is - without it, items disappear after the flipmap. So that comment is removed.

Things to test:
- general flipmap activation/deactivation
- saves in various flip states
- flip effects, like rushing water in The Cistern
- built-in flipmaps: lightning in Folly, Pyramid; cabin in Natla's Mines; lava bowl in Ice Palace
- dead enemies in flipped rooms in TR2 should disappear (like Living Quarters)
- pushblocks should still work, like Barkhang room 150 after flipping

lahm86 committed:
room: move flip status to TRX

This moves the global flip status flag to TRX.

lahm86 committed:
room: move flip effect to TRX

This moves the global flip effect variable to TRX.

lahm86 committed:
room: move flip timer to TRX

This moves the global flip timer variable to TRX.

lahm86 committed:
room: move flip slot flags to TRX

This moves the global flip slot flags table to TRX.

lahm86 committed:
room: move flipmap logic to TRX

This moves the main flipmap logic itself to TRX.

rr- commented on src/tr1/game/item_actions/chain_block.c:
```diff
@@ -8,20 +8,21 @@
 
-    g_FlipTimer++;
-    if (g_FlipTimer == 55) {
+    Room_SetFlipTimer(++flip_timer);
```
Let's avoid preincrementing the arguments.

lahm86 replied:
Fixed, coupled with update below.

rr- commented on src/tr1/game/item_actions/flood.c:
```diff
@@ -23,5 +24,5 @@ void ItemAction_Flood(ITEM *item)
     }
 
-    g_FlipTimer++;
+    Room_SetFlipTimer(++flip_timer);
```
Perhaps Room_IncrementTimer(int32_t frames);?

lahm86 replied:
Thanks, I've introduced this and tidied things up. In some cases we check frame numbers so that's been tweaked to take into account the local copies of the value, which are also now const for clarity.

rr- commented on src/tr1/game/level.c:
```diff
@@ -594,15 +594,9 @@ bool Level_Initialise(const GF_LEVEL *const level)
-    for (int32_t i = 0; i < MAX_FLIP_MAPS; i++) {
-        g_FlipMapTable[i] = 0;
-    }
-
```
Is it safe to remove this?

lahm86 replied:
Yes, this has been merged into `Room_InitialiseFlipStatus`, which we call from `Level_ReadRooms`. So it just happens a little later, but still before the level actually finishes loading.

rr- commented on src/tr1/game/savegame/savegame_legacy.c:
```diff
@@ -665,8 +665,12 @@ bool Savegame_Legacy_LoadFromFile(MYFILE *fp, GAME_INFO *game_info)
-    M_Read(&g_FlipTimer, sizeof(int32_t));
+    int32_t flip_effect;
+    int32_t flip_timer;
+    M_Read(&flip_effect, sizeof(int32_t));
```
Let's use M_ReadS32 as a safer alternative.

lahm86 replied:
We don't have this in TR1's legacy save module. Should we introduce it from TR2?

rr- replied:
Ideally, yes,  but since this would be about changing the entire file, let's do that separately.

rr- reviewed (commented):
Thank you 😌

lahm86 committed:
room: introduce timer incrementer

This avoids having to locally adjust the flip timer and then set it,
by using a function to directly add frames.
